helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/IZkStateListener.java

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright 2010 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.helix.manager.zk.zookeeper;
+
+import org.apache.zookeeper.Watcher.Event.KeeperState;
+
+public interface IZkStateListener {
+
+  /**
+   * Called when the zookeeper connection state has changed.
+   *
+   * @param state
+   *            The new state.
+   * @throws Exception
+   *             On any error.
+   */
+  void handleStateChanged(KeeperState state) throws Exception;
+
+  /**
+   * Called after the zookeeper session has expired and a new session has been created.
+   * You would have to re-create any ephemeral nodes here.
+   *
+   * @deprecated
+   * This is deprecated, because it may cause session race condition when creating ephemeral nodes.
+   * Please use {@link #handleNewSession(String)} instead, which takes care of race condition.
+   *
+   * @throws Exception
+   *             On any error.
+   */
+  @Deprecated
+  void handleNewSession() throws Exception;
+
+  /**
+   * Called after the zookeeper session has expired and a new session has been created. The new
+   * session id has to be passed in as the parameter. And you would have to re-create any ephemeral
+   * nodes here.
```

Comment:
Explain a little more about why you'd have to ephemeral nodes here? That's just Helix-specific logic right?

Could a user have a different set of logic in their `handleNewSession` implementation?

I appreciate the comment because I think it serves as a helpful example for usage though. Let's just rephrase this JavaDoc.

Reply:
This interface is referred to I0Itec's. I just did not want to change the implementation logic in the comment. My understanding is, this is an operation to handle a new session. The ephemeral nodes in the expired session are gone. So in this handle new session operation, new ephemeral nodes are expected to create. Correct me if I am wrong.